File: src/schemas/metadata/MetadataTemplate.ts
import { ClassProperties } from "../../typings/ClassProperties";

import { IMetadataDictionary } from "./IMetadataDictionary";
import { IMetadataTemplate } from "./IMetadataTemplate";
import { IMetadataTypeTag } from "./IMetadataTypeTag";
import { Metadata } from "./Metadata";

export class MetadataTemplate {
  public readonly row: Metadata[];
  public readonly tags: IMetadataTypeTag[][];

  private name_?: string;

  private constructor(props: ClassProperties<MetadataTemplate>) {
    this.row = props.row.map(Metadata.create);
    this.tags = props.tags;
  }

  public static create(
    props: ClassProperties<MetadataTemplate>,
  ): MetadataTemplate {
    return new MetadataTemplate(props);
  }

  public static from(
    json: IMetadataTemplate,
    dict: IMetadataDictionary,
  ): MetadataTemplate {
    return new MetadataTemplate({
      row: json.row.map((m) => Metadata.from(m, dict)),
      tags: json.tags,
    });
  }

  public getName(): string {
    return (this.name_ ??= getName(this));
  }

  /**
   * @internal
   */
  public getBaseName(): string {
    return (
      "`" +
      this.row
        .map((child) =>
          child.isConstant() && child.size() === 1
            ? child.constants[0]!.values[0]!.value
            : `$\{${child.getName()}\}`,
        )
        .join("")
        .split("`")
        .join("\\`") +
      "`"
    );
  }

  public toJSON(): IMetadataTemplate {
    return {
      row: this.row.map((m) => m.toJSON()),
      tags: this.tags,
    };
  }
}

const getName = (template: MetadataTemplate): string => {
  const base: string = template.getBaseName();
  if (!template.tags?.length) return base;
  else if (template.tags.length === 1) {
    const str: string = [base, ...template.tags[0]!.map((t) => t.name)].join(
      " & ",
    );
    return `(${str})`;
  }
  const rows: string[] = template.tags.map((row) => {
    const str: string = row.map((t) => t.name).join(" & ");
    return row.length === 1 ? str : `(${str})`;
  });
  return `(${base} & (${rows.join(" | ")}))`;
};

File: packages/typia/native/core/schemas/metadata/MetadataTemplate.go
package metadata

import "strings"

type IMetadataSchema_ITemplate struct {
  Row  []*IMetadataSchema
  Tags [][]IMetadataTypeTag
}

type MetadataTemplate struct {
  Row   []*MetadataSchema
  Tags  [][]IMetadataTypeTag
  name_ string
}

func MetadataTemplate_create(props MetadataTemplate) *MetadataTemplate {
  row := make([]*MetadataSchema, 0, len(props.Row))
  for _, child := range props.Row {
    row = append(row, MetadataSchema_create(*child))
  }
  return &MetadataTemplate{
    Row:  row,
    Tags: cloneTagMatrix(props.Tags),
  }
}

func MetadataTemplate_from(json IMetadataSchema_ITemplate, dict IMetadataDictionary) *MetadataTemplate {
  row := make([]*MetadataSchema, 0, len(json.Row))
  for _, elem := range json.Row {
    row = append(row, MetadataSchema_from(elem, dict))
  }
  return &MetadataTemplate{
    Row:  row,
    Tags: cloneTagMatrix(json.Tags),
  }
}

func (obj *MetadataTemplate) GetName() string {
  if obj.name_ == "" {
    obj.name_ = metadataTemplate_getName(obj)
  }
  return obj.name_
}

func (obj *MetadataTemplate) GetBaseName() string {
  parts := make([]string, 0, len(obj.Row))
  for _, child := range obj.Row {
    if child.IsConstant() && child.Size() == 1 {
      parts = append(parts, child.Constants[0].Values[0].Value.(string))
    } else {
      parts = append(parts, "${"+child.GetName()+"}")
    }
  }
  return "`" + strings.ReplaceAll(strings.Join(parts, ""), "`", "\\`") + "`"
}

func (obj *MetadataTemplate) ToJSON() IMetadataSchema_ITemplate {
  row := make([]*IMetadataSchema, 0, len(obj.Row))
  for _, elem := range obj.Row {
    row = append(row, elem.ToJSON())
  }
  return IMetadataSchema_ITemplate{
    Row:  row,
    Tags: cloneTagMatrix(obj.Tags),
  }
}

func metadataTemplate_getName(template *MetadataTemplate) string {
  return taggedName(template.GetBaseName(), template.Tags)
}
